x-pack/plugins/beats/server/routes/api/register_remove_tags_from_beats_route.js

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import Joi from 'joi';
+import {
+  get,
+  flatten,
+  uniq
+} from 'lodash';
+import { INDEX_NAMES } from '../../../common/constants';
+import { callWithRequestFactory } from '../../lib/client';
+import { wrapEsError } from '../../lib/error_wrappers';
+
+async function getDocs(callWithRequest, ids) {
+  const params = {
+    index: INDEX_NAMES.BEATS,
+    type: '_doc',
+    body: { ids },
+    _source: false
+  };
+
+  const response = await callWithRequest('mget', params);
+  return get(response, 'docs', []);
+}
+
+function getBeats(callWithRequest, beatIds) {
+  const ids = beatIds.map(beatId => `beat:${beatId}`);
+  return getDocs(callWithRequest, ids);
+}
+
+function getTags(callWithRequest, tags) {
+  const ids = tags.map(tag => `tag:${tag}`);
+  return getDocs(callWithRequest, ids);
+}
+
+async function findNonExistentItems(callWithRequest, items, getFn) {
+  const itemsFromEs = await getFn.call(null, callWithRequest, items);
+  return itemsFromEs.reduce((nonExistentItems, itemFromEs, idx) => {
+    if (!itemFromEs.found) {
+      nonExistentItems.push(items[idx]);
+    }
+    return nonExistentItems;
+  }, []);
+}
+
+function findNonExistentBeatIds(callWithRequest, beatIds) {
+  return findNonExistentItems(callWithRequest, beatIds, getBeats);
+}
+
+function findNonExistentTags(callWithRequest, tags) {
+  return findNonExistentItems(callWithRequest, tags, getTags);
+}
+
+async function persistRemovals(callWithRequest, removals) {
+  const body = flatten(removals.map(removal => {
+    const { beatId, tag } = removal;
+    const script = 'def beat = ctx._source.beat; '
+      + 'if (beat.tags == null) { '
+      + '  beat.tags = []; '
+      + '} '
+      + 'beat.tags.removeAll([params.tag]);';
```

Comment:
@nik9000 Do you have any suggestions for this script? I'm using it in a bulk API call, doc update operation (see a couple lines below here where I construct the bulk API request body "rows").

My docs in ES have a field called `beat`, which has a sub-field called `tags`. Here I'm trying to remove a tag, if it exists, from the list of `tags`.

Reply:
I'd probably skip removing anything if the array doesn't exist.

I'd probably do the whole `indexOf` thing and then `remove` it rather than a list. I wonder why we haven't found a way to whitelist the object variant of `remove`..... If this list gets long you could probably be more efficient by removing the last element in the list and putting it on top of the element that you actually want to remove. Then we don't have to shift everything back on position. But that is harder to read and easier to make a mistake while doing so it isn't worth it if the list doesn't get very big.

Reply:
> I'd probably skip removing anything if the array doesn't exist.

Ha, yes of course, duh! Thanks for pointing this out.

> I'd probably do the whole indexOf thing and then remove it rather than a list. 

Sounds good, but why is this better (asking more for my own curiosity at this point)? Sounds like `indexOf` would be `O(n)`, as would what I have now, presumably?

> If this list gets long you could probably be more efficient by removing the last element in the list and putting it on top of the element that you actually want to remove. Then we don't have to shift everything back on position. But that is harder to read and easier to make a mistake while doing so it isn't worth it if the list doesn't get very big.

Nice! I don't expect these lists to get more than dozens of tags long, so yeah I'm not sure if the readability tradeoff is worth the performance gain. But this is a good option to have in our back pocket. Thanks!